### PR TITLE
[Model] Phi-3 4k sliding window temp. fix

### PR DESCRIPTION
**Pull request by caiom**
The model microsoft/Phi-3-mini-4k-instruct uses a sliding window of size 2047 (which should result in 2048 tokens of context) but due to the issue #3385 the model loading will break:

`File "/usr/local/lib/python3.10/dist-packages/vllm/core/block_manager_v1.py", line 223, in __init__
    assert sliding_window % block_size == 0, (sliding_window,
AssertionError: (2047, 16)`

My idea is to have this quick fix so that people can use the model, I can also create a draft PR for a definitive solution. However, a definitive solution to the issue can be involved, the issue affects xformers and sdpa but not flash backend. It also affects _block_manager_v1_ which is the bit I would need some help.

Let me know your thoughts.

RELATED: https://github.com/vllm-project/vllm/issues/3385

**Update:**

After some discussion, we decided to change only `block_manager_v1` by removing the assert and rounding up `block_sliding_window`. This change has no effect anywhere else (`model_runner` or actual model code).

Models tested by feeding long prompts: 

- microsoft/Phi-3-mini-4k-instruct
- mistralai/Mistral-7B-Instruct-v0.1

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by esmeetu** [comment]
@caiom Should we apply this padding in `llama.py`?

**Comment by caiom** [comment]
> @caiom Should we apply this padding in `llama.py`?

For a window size of 2047 we have:

1. xformers will use 2047 tokens -> This is fine since it is one less than the correct.
2. sdpa will use 2047 tokens -> This is fine since it is one less than the correct.
3. flash will use 2048 tokens -> Correct.

This is true for all Models that uses a sliding window.

With this bypass, we are just avoiding the assert in _block_manager_v1_. But for Mistral, the _block_manager_v1_ will think the window size is smaller than it actually is (with the flash backend). The window size of mistralai/Mistral-7B-v0.1 is 4096 and hence the flash backend will mask 4097 tokens, while I guess _block_manager_v1_ is assuming 4096 tokens.

For Phi-3, with this PR, _block_manager_v1_ may assume a larger number of tokens than the actual number. For instance, with the window size of 2047, the PR will add one and hence the _block_manager_v1_ will see it as 2048 tokens. However, with sdpa that number may be 2047. 

So, there's mismatch between _block_manager_v1_ and the actual window size. This is true for all Models, for more or for less and depending on the backend.

Now, I'm not sure if any of these cases of mismatch are a significant problem as I don't really know how _block_manager_v1_ actually works.

Hopefully my explanation is good enough :)

**Comment by esmeetu** [comment]
@caiom Thanks for your detail explanation. Look good to me. @cadedaniel WDYT?

**Review by esmeetu — APPROVED** [review]

**Comment by cadedaniel** [comment]
What if instead we round up the sliding window in the block manager to nearest block size?

**Comment by caiom** [comment]
I'm guessing that for PagedAttention, what matters is what is in the blocks_table. The max context length for mistralai/Mistral-7B-v0.1 is 4097 but the PagedAttention will consider only 4096 tokens. This is fine since the model can handle 4096 tokens.

For Phi-3, using this PR, PagedAttention have the correct/maximum number of tokens in context. 

So, if we are going to round it, we should always round down so that we never feed more tokens than the model was trained for. Unless we are just rounding by 1, so the +1 in my PR should be safe.

```model_runner.py``` will round down as it is. It is better to have it consistent with ```block_manager_v1``` so that is does not PAD more tokens than needed.

I may be totally wrong, though.

**Comment by cadedaniel** [comment]
So the block manager's concern here (and only concern) should be allocating enough space for sliding window. As for how the model would like to use that space is up to the model. So from a north-star perspective the block manager should not be involved in the minutia of context len for sliding window. It's OK if the model_runner uses a subset of the blocks for context.

While this PR is a working hotfix, in design terms it couples the exact sliding window length used by attention closely with the block manager, which isn't necessary. Ideally we avoid this!

Can we test out my suggestion? There may be some other coupling between block manager and exact context length that I don't know about.

**Comment by caiom** [comment]
Thanks for the review @cadedaniel, I have implemented your suggestion and tested _microsoft/Phi-3-mini-4k-instruct_ using several long prompts, everything looks good.

File: vllm/core/block_manager_v1.py [hunk 1]
```diff
@@ -220,9 +221,7 @@ def __init__(
 
         self.block_sliding_window = None
         if sliding_window is not None:
-            assert sliding_window % block_size == 0, (sliding_window,
-                                                      block_size)
-            self.block_sliding_window = sliding_window // block_size
+            self.block_sliding_window = math.ceil(sliding_window / block_size)
```
Comment (cadedaniel) [inline vllm/core/block_manager_v1.py:224]: Can we add a comment with the motivation for `ceil`? suggestion:
```
# Round to nearest block size to regularize sliding window allocation sizes.
```

**Review by cadedaniel — COMMENTED** [review]
LGTM, thanks! Small thing:
* can you update the PR description with latest?
* can you run a manual test with `mistralai/Mistral-7B-v0.1`? I tried but my node is borked at the moment..

**Comment by caiom** [comment]
Done! I tested the _mistralai/Mistral-7B-Instruct-v0.1_ model by inputting Wikipedia articles, asking it to create summaries, and then eyeballing the results. Everything looks good. Additionally, the rounding should not affect Mistral models.

**Review by cadedaniel — APPROVED** [review]
Thanks for the fix! We can merge once CI finishes its run

**Comment by ShadowTeamCN** [comment]
> Thanks for the review @cadedaniel, I have implemented your suggestion and tested _microsoft/Phi-3-mini-4k-instruct_ using several long prompts, everything looks good.

hi, I find Phi-3 medium is released ,when i try it with vllm, framework logged it was using xformers because of sliding window, my vllm is building from latest source, do you have any idea how to use flash-attn instead of xformers?

> INFO 05-22 11:09:12 selector.py:116] Cannot use FlashAttention-2 backend due to sliding window.
> INFO 05-22 11:09:12 selector.py:52] Using XFormers backend.

**Comment by alexanderfrey** [comment]
@ShadowTeamCN Same problem here for phi3-mini. Cannot use FlashAttention-2 backend due to sliding window.

**Comment by davidgxue** [comment]
Hey wanted to check: is this problem fixed? Both my mistral and phi 3 model cannot use FA2
```
INFO 07-26 19:19:35 selector.py:170] Cannot use FlashAttention-2 backend due to sliding window.
INFO 07-26 19:19:35 selector.py:54] Using XFormers backend.
```
and Im using the latest version of vLLM

**Comment by ryancurrah** [comment]
I also an unable to use flash attention with `microsoft/Phi-3.5-MoE-instruct`.

**Comment by makramhamdaoui** [comment]
Same problem here for Phi-3.5-vision. Cannot use FlashAttention-2 backend due to sliding window.

**Pull request merged**